[CoSim] New data transfer operator

RiedlAndreas opened the pull request:
**📝 Description**
The new data transfer operator (SumDistributedToDistributed), sums distributed values from one interface and transfers them to another interface. If desired equally distributed. The interfaces do not have to be the same size or located at the same position.
Already able to perform in MPI.

I used it to transfer Reaction forces from a CFD model to a different interface in the CSD model.

**🆕 Changelog**
- Added new data transfer operator

RiedlAndreas committed:
New data transfer operation
sum distributed to distributed

philbucher commented on applications/CoSimulationApplication/python_scripts/data_transfer_operators/sum_distributed_to_distributed.py:
```diff
@@ -0,0 +1,43 @@
+def Create(settings):
+    return SumDistributedToDistributed(settings)
+
+class SumDistributedToDistributed(CoSimulationDataTransferOperator):
```
```suggestion
class SumMultipleToMultiple(CoSimulationDataTransferOperator):
```

I also want to rename the other classes, since `distributed` is associated with MPI in Kratos

philbucher commented on applications/CoSimulationApplication/python_scripts/data_transfer_operators/sum_distributed_to_distributed.py:
```diff
@@ -0,0 +1,43 @@
+    """
+    def _ExecuteTransferData(self, from_solver_data, to_solver_data, transfer_options):        
+        if not to_solver_data.IsDefinedOnThisRank():
+            return
```
this is wrong

Imagine `from_solver` is using this rank then you deadlock when doing `SumAll`

you have to do some more checks, see the other classes

philbucher commented on applications/CoSimulationApplication/python_scripts/data_transfer_operators/sum_distributed_to_distributed.py:
```diff
@@ -0,0 +1,43 @@
+        value = from_solver_data.GetModelPart().GetCommunicator().GetDataCommunicator().SumAll(value)     
+
+        to_solver_values = to_solver_data.GetData()
+        to_solver_values.fill(value)
```
are you sure this is correct?
This would increase the total amount of whatever you exchange

On one side you sum all, then on the other side you set this value to EVERY entity

RiedlAndreas replied:
For my case I used the transfer_operator "distribute_values" to distribute the sum equally to all entities.

In this case the sum of both sides is equal.

philbucher replied:
ok thats what I thought

Can you please explain this clearly in the docstring to avoid confusion?

philbucher reviewed (commented):
some comments and please add a [test](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/CoSimulationApplication/tests/test_data_transfer_operators.py)

RiedlAndreas committed:
Merge branch 'master' into cosim/data_transfer_mpi

RiedlAndreas committed:
Add data_transfer_operator with new name

RiedlAndreas committed:
No sum if from_solver is not distributed

RiedlAndreas committed:
New tests of sum many to many

-Change all data transfer operators to many_to_many
- New scenario with many_to_non_matching_many

RiedlAndreas committed:
Update test_data_transfer_operators for mpi

MPI tests
